src/python/pants/backend/python/macros/python_requirements.py

```diff
@@ -69,6 +70,11 @@ def __call__(
         requirements = parse_requirements_file(
             req_file.read_text(), rel_path=str(req_file.relative_to(get_buildroot()))
         )
+
+        if module_mapping is None:
+            module_mapping = self._global_module_mapping
+        # TODO: else merge?
```

Comment:
I do think that merging would make sense here, as otherwise there is a cliff where if you need to define any new module that isn't in the defaults, you suddenly have to re-define all of them.

But merging is challenging from a backwards compatibility perspective: if a future version of Pants adds additional mappings, it would break inference for older code by introducing collisions.

That leaves me unsure as to whether Pants should build in this file. It might be workable to expose an option that a user could use to set a YAML file containing an expanded mapping (which would be stable across versions?), or to publish an independent artifact containing the mapping, so that that could change independently from your pants version?

Reply:
Ah, tricky. I'll keep this in mind, going forward see if I get any ideas how one might tackle it.

Reply:
As decided in the main thread, we must always merge the two, but with the user-provided mapping winning out every time on conflicts.

We do not plan to expose an option to change the defaults: the way you do that is overriding it via the macro/target.